R/render-cached-plot.R

```diff
@@ -0,0 +1,404 @@
+#'
+#' @export
+renderCachedPlot <- function(expr, cacheKeyExpr, cacheResetEventExpr = NULL,
+  baseWidth = 400, aspectRatioRate = 1.2, growthRate = 1.2, res = 72,
+  scope = "app",
+  env = parent.frame(), quoted = FALSE, outputArgs = list()
+) {
+
+  # This ..stacktraceon is matched by a ..stacktraceoff.. when plotFunc
+  # is called
+  installExprFunction(expr, "func", env, quoted, ..stacktraceon = TRUE)
+  # This is so that the expr doesn't re-execute by itself; it needs to be
+  # triggered by the cache key (or width/height) changing.
+  isolatedFunc <- function() isolate(func())
+
+  cacheKey <- reactive(substitute(cacheKeyExpr), env = parent.frame(), quoted = TRUE)
+
+  .cacheDir <- NULL
+  cacheDir <- function() {
+    # Memoize
+    if (is.null(.cacheDir)) {
+      if (is.null(outputName)) {
+        stop("outputName is NULL. cacheDir() was called too early.")
+      }
+
+      if (scope %in% c("app", "session")) {
+        appCachePath <- file.path(tempdir(), paste0("shinyapp-", getShinyOption("appToken")))
+
+        if (scope == "app") {
+          cacheScopePath <- appCachePath
+        } else if (scope == "session") {
+          cacheScopePath <- file.path(appCachePath, paste0("shinysession-", session$token))
+        }
+
+        .cacheDir <<- file.path(cacheScopePath, paste0("output-", outputName))
+
+      } else {
+        # User has passed in a directory
+        .cacheDir <<- normalizePath2(scope)
+      }
+    }
+
+    .cacheDir
+  }
+
+  ensureCacheDirExists <- function() {
+    if (!dirExists(cacheDir())) {
+      cat("Creating ", cacheDir(), "\n")
+      dir.create(cacheDir(), recursive = TRUE, mode = "0700")
+
+      # Set up removal of cache directory at appropriate time. The removal
+      # callback is registered here, paired with the creation of the cache
+      # dir, to ensure it's not scheduled multiple times for one directory.
+      deleteCacheDir <- function() {
+        # Just to be safe, don't try to delete the cache dir if it's already
+        # gone.
+        if (!dirExists(cacheDir())) {
+          return()
+        }
+
+        unlink(cacheDir(), recursive = TRUE)
+
+        # Recursively delete empty parent dirs, up to temp dir.
+        currentDir <- dirname(cacheDir())
+        while (currentDir != tempdir() &&
+               length(dir(currentDir, all.files = TRUE, no.. = TRUE)) == 0)
+        {
+          dirRemove(currentDir)
+          currentDir <- dirname(currentDir)
+        }
+      }
+
+      if (scope == "app") {
+        onStop(deleteCacheDir, session = NULL)
+      } else if (scope == "session") {
+        onSessionEnded(deleteCacheDir)
+      }
+    }
+  }
+
+  # Clear the cacheDir at the appropriate time. Use ignoreInit=TRUE because we
+  # don't want it to happen right in the beginning.
+  observeEvent(
+    substitute(cacheResetEventExpr), event.env = parent.frame(), event.quoted = TRUE,
+    ignoreInit = TRUE,
+    {
+      unlink(file.path(cacheDir(), "*.rds"))
```

Comment:
Make safer for multiple processes

Reply:
I seem to recall that this had something to do with the cache possibly being reset at the start of a new session, but I can't remember the details.